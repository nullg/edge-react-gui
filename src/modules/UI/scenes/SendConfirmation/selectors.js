// @flow

import type { EdgeMetadata, EdgeSpendInfo, EdgeTransaction } from 'edge-core-js'

import { STANDARD_FEE } from '../../../../constants/indexConstants'
import { type GuiMakeSpendInfo } from '../../../../reducers/scenes/SendConfirmationReducer.js'
import type { State } from '../../../ReduxTypes'
import { convertNativeToExchange } from '../../../utils.js'
import { convertCurrency, getSceneState, getSelectedCurrencyCode } from '../../selectors.js'
import { getExchangeDenomination } from '../../Settings/selectors.js'

export const initialState = {
  isKeyboardVisible: false,
  forceUpdateGuiCounter: 0,

  parsedUri: {
    networkFeeOption: (STANDARD_FEE: string),
    customNetworkFee: {},
    publicAddress: '',
    nativeAmount: '0',
    metadata: {
      name: '',
      category: '',
      notes: '',
      amountFiat: 0,
      bizId: 0,
      miscJson: ''
    }
  },
  spendInfo: null,
  transactionMetadata: null,
  nativeAmount: '0',

  isEditable: true,

  transaction: {
    txid: '',
    date: 0,
    currencyCode: '',
    blockHeight: -1,
    nativeAmount: '0',
    networkFee: '',
    parentNetworkFee: '',
    ourReceiveAddresses: [],
    signedTx: '',
    metadata: {},
    otherParams: {}
  },
  pending: false,
  error: null,

  pin: '',
  authRequired: 'none',
  address: ''
}

export const getScene = (state: State): any => getSceneState(state, 'sendConfirmation')
export const getPending = (state: State): boolean => getScene(state).pending
export const getError = (state: State): Error => getScene(state).error
export const getKeyboardIsVisible = (state: State): boolean => getScene(state).keyboardIsVisible

export const getTransaction = (state: State): EdgeTransaction => getScene(state).transaction || initialState.transaction
export const getParsedUri = (state: State): GuiMakeSpendInfo => getScene(state).parsedUri || initialState.parsedUri
export const getForceUpdateGuiCounter = (state: State): number => getScene(state).forceUpdateGuiCounter

export const getNetworkFeeOption = (state: State): string => getParsedUri(state).networkFeeOption || initialState.parsedUri.networkFeeOption || ''
export const getCustomNetworkFee = (state: State): any => getParsedUri(state).customNetworkFee || initialState.parsedUri.customNetworkFee || {}
export const getMetadata = (state: State): EdgeMetadata => getParsedUri(state).metadata || initialState.parsedUri.metadata || {}
export const getPublicAddress = (state: State): string => {
  try {
    return (
      getParsedUri(state).publicAddress ||
      initialState.parsedUri.publicAddress ||
      // $FlowFixMe
      state.ui.scenes.sendConfirmation.spendInfo.spendTargets[0].publicAddress ||
      ''
    )
  } catch (e) {
    return ''
  }
}
export const getNativeAmount = (state: State): string | void => state.ui.scenes.sendConfirmation.nativeAmount

export const getUniqueIdentifier = (state: State): string => {
  const parsedUri = getParsedUri(state)
  const uniqueIdentifier = parsedUri.uniqueIdentifier
  return uniqueIdentifier || ''
}

export const getNetworkFee = (state: State): string => getTransaction(state).networkFee
export const getParentNetworkFee = (state: State): string | void => getTransaction(state).parentNetworkFee

export const getSpendInfo = (state: State, newSpendInfo?: GuiMakeSpendInfo = {}): EdgeSpendInfo => {
  const uniqueIdentifier = newSpendInfo.uniqueIdentifier || getUniqueIdentifier(state)

  return {
    currencyCode: newSpendInfo.currencyCode || getSelectedCurrencyCode(state),
    metadata: newSpendInfo.metadata ? { ...getMetadata(state), ...newSpendInfo.metadata } : getMetadata(state),
    spendTargets: [
      {
        nativeAmount: newSpendInfo.nativeAmount || getNativeAmount(state),
        publicAddress: newSpendInfo.publicAddress || getPublicAddress(state),
        otherParams: {
          uniqueIdentifier
        }
      }
    ],
    networkFeeOption: newSpendInfo.networkFeeOption || getNetworkFeeOption(state),
    customNetworkFee: newSpendInfo.customNetworkFee ? { ...getCustomNetworkFee(state), ...newSpendInfo.customNetworkFee } : getCustomNetworkFee(state)
  }
}

export type AuthType = 'pin' | 'none'
export const getAuthRequired = (state: State, spendInfo: EdgeSpendInfo): AuthType => {
  const isEnabled = state.ui.settings.spendingLimits.transaction.isEnabled
  if (!isEnabled) return 'none'

  const currencyCode = spendInfo.currencyCode || spendInfo.spendTargets[0].currencyCode
  const { nativeAmount } = spendInfo.spendTargets[0]
  if (!currencyCode || !nativeAmount) throw new Error('Invalid Spend Request')

  const { spendingLimits } = state.ui.settings
  const isoFiatCurrencyCode = state.ui.settings.defaultIsoFiat
  const nativeToExchangeRatio = getExchangeDenomination(state, currencyCode).multiplier
  const exchangeAmount = convertNativeToExchange(nativeToExchangeRatio)(nativeAmount)
  const fiatAmount = convertCurrency(state, currencyCode, isoFiatCurrencyCode, parseFloat(exchangeAmount))
  const exceedsLimit = fiatAmount >= spendingLimits.transaction.amount

  return exceedsLimit ? 'pin' : 'none'
}
