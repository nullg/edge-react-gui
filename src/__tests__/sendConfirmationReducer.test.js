/* eslint-disable flowtype/require-valid-file-annotation */
/* globals describe test expect */

import { clone } from 'ramda'

import {
  makeSpendFailed,
  newPin,
  newSpendInfo,
  reset,
  updatePaymentProtocolTransaction,
  updateSpendPending,
  updateTransaction
} from '../actions/SendConfirmationActions.js'
import { initialState } from '../modules/UI/scenes/SendConfirmation/selectors.js'
import { sendConfirmation } from '../reducers/scenes/SendConfirmationReducer.js'

describe('sendConfirmation reducer', () => {
  test('initialState', () => {
    const actual = sendConfirmation(undefined, {})

    expect(actual).toMatchSnapshot()
  })

  test('reset', () => {
    const action = reset()
    const actual = sendConfirmation(undefined, action)

    expect(actual).toMatchSnapshot()
  })

  describe('destination', () => {
    describe('updateTransaction', () => {
      test('with transaction and legacyAddress', () => {
        const parsedUri = {
          publicAddress: 'bitcoincash:qpltjkre069mp80ylcj87832ju3zt2gr6gercn9j2z',
          legacyAddress: '123412341234',
          nativeAmount: '100000',
          currencyCode: 'BCH',
          metadata: {}
        }
        const transaction = {
          blockHeight: 0,
          currencyCode: 'BCH',
          date: 0,
          nativeAmount: '-681',
          networkFee: '681',
          otherParams: {},
          ourReceiveAddresses: ['123123123'],
          signedTx: '',
          txid: ''
        }
        const error = null
        const forceUpdateGui = true
        const initialStateClone = clone(initialState)
        const action = updateTransaction(transaction, parsedUri, forceUpdateGui, error)
        const actual = sendConfirmation(initialStateClone, action) // use initialState after sendConfirmation reducer not longer mutates state

        expect(actual).toMatchSnapshot()
      })

      test('with transaction and name', () => {
        const parsedUri = {
          publicAddress: 'bitcoincash:qpltjkre069mp80ylcj87832ju3zt2gr6gercn9j2z',
          nativeAmount: '100000',
          currencyCode: 'BCH',
          metadata: {
            name: 'airbitz'
          }
        }
        const transaction = {
          blockHeight: 0,
          currencyCode: 'BCH',
          date: 0,
          nativeAmount: '-681',
          networkFee: '681',
          otherParams: {},
          ourReceiveAddresses: ['123123123'],
          signedTx: '',
          txid: ''
        }
        const error = null
        const forceUpdateGui = true
        const initialStateClone = clone(initialState)
        const action = updateTransaction(transaction, parsedUri, forceUpdateGui, error)
        const actual = sendConfirmation(initialStateClone, action) // use initialState after sendConfirmation reducer not longer mutates state

        expect(actual).toMatchSnapshot()
      })

      test('with error', () => {
        const transaction = null
        const parsedUri = { nativeAmount: '100000' }
        const forceUpdateGui = true
        const error = new Error()
        const initialStateClone = clone(initialState)
        const action = updateTransaction(transaction, parsedUri, forceUpdateGui, error)
        const actual = sendConfirmation(initialStateClone, action) // use initialState after sendConfirmation reducer not longer mutates state

        expect(actual).toMatchSnapshot()
      })

      test('with pin error', () => {
        const parsedUri = null
        const transaction = {
          blockHeight: 0,
          currencyCode: 'BCH',
          date: 0,
          nativeAmount: '-681',
          networkFee: '681',
          otherParams: {},
          ourReceiveAddresses: ['123123123'],
          signedTx: '',
          txid: ''
        }
        const error = new Error('Incorrect Pin')
        const forceUpdateGui = true
        const initialStateClone = clone(initialState)
        const action = updateTransaction(transaction, parsedUri, forceUpdateGui, error)
        const actual = sendConfirmation(initialStateClone, action) // use initialState after sendConfirmation reducer not longer mutates state

        expect(actual).toMatchSnapshot()
      })
    })

    describe('newSpendInfo', () => {
      test('with name', () => {
        const spendInfo = {
          currencyCode: 'BTC',
          nativeAmount: '1000',
          spendTargets: [{ currencyCode: 'BTC', nativeAmount: '1000', publicAddress: '123123123' }],
          metadata: { name: 'airbitz' }
        }
        const action = newSpendInfo(spendInfo, 'none')
        const initialStateClone = clone(initialState)
        const actual = sendConfirmation(initialStateClone, action)

        expect(actual).toMatchSnapshot()
      })

      test('without name', () => {
        const spendInfo = {
          currencyCode: 'BTC',
          nativeAmount: '1000',
          spendTargets: [{ currencyCode: 'BTC', nativeAmount: '1000', publicAddress: '123123123' }],
          metadata: {}
        }
        const action = newSpendInfo(spendInfo, 'none')
        const initialStateClone = clone(initialState)
        const actual = sendConfirmation(initialStateClone, action)

        expect(actual).toMatchSnapshot()
      })
    })
  })

  describe('address', () => {
    test('NEW_SPEND_INFO', () => {
      const spendInfo = {
        spendTargets: [{ publicAddress: '123123123', nativeAmount: '0' }],
        metadata: {}
      }
      const action = newSpendInfo(spendInfo, 'none')
      const initialStateClone = clone(initialState)

      const actual = sendConfirmation(initialStateClone, action)

      expect(actual).toMatchSnapshot()
    })
  })

  describe('isEditable', () => {
    test('UPDATE_PAYMENT_PROTOCOL_TRANSACTION', () => {
      const edgeTransaction = { id: '123', nativeAmount: '123' }
      const action = updatePaymentProtocolTransaction(edgeTransaction)
      const initialStateClone = clone(initialState)
      const actual = sendConfirmation(initialStateClone, action)

      expect(actual).toMatchSnapshot()
    })

    test('MAKE_PAYMENT_PROTOCOL_TRANSACTION_FAILED', () => {
      const error = new Error()
      const action = makeSpendFailed(error)
      const initialStateClone = clone(initialState)
      const actual = sendConfirmation(initialStateClone, action)

      expect(actual).toMatchSnapshot()
    })
  })

  describe('error', () => {
    test('UPDATE_TRANSACTION', () => {
      const error = new Error()
      const action = updateTransaction(null, null, null, error)
      const initialStateClone = clone(initialState)
      const actual = sendConfirmation(initialStateClone, action)

      expect(actual).toMatchSnapshot()
    })

    test('MAKE_PAYMENT_PROTOCOL_TRANSACTION_FAILED', () => {
      const error = new Error()
      const action = makeSpendFailed(error)
      const initialStateClone = clone(initialState)
      const actual = sendConfirmation(initialStateClone, action)

      expect(actual).toMatchSnapshot()
    })
  })

  test('pin', () => {
    const action = newPin('1234')
    const initialStateClone = clone(initialState)
    const actual = sendConfirmation(initialStateClone, action)

    expect(actual).toMatchSnapshot()
  })

  test('pending', () => {
    const action = updateSpendPending(true)
    const initialStateClone = clone(initialState)
    const actual = sendConfirmation(initialStateClone, action)

    expect(actual).toMatchSnapshot()
  })
})
