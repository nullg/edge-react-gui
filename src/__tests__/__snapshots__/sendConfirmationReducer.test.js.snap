// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendConfirmation reducer address NEW_SPEND_INFO 1`] = `
Object {
  "address": "123123123",
  "authRequired": "none",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "123123123",
  },
  "pending": false,
  "pin": "",
  "spendInfo": Object {
    "metadata": Object {},
    "spendTargets": Array [
      Object {
        "nativeAmount": "0",
        "publicAddress": "123123123",
      },
    ],
  },
  "transaction": null,
  "transactionMetadata": Object {},
}
`;

exports[`sendConfirmation reducer destination newSpendInfo with name 1`] = `
Object {
  "address": "123123123",
  "authRequired": "none",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "1000",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "airbitz",
      "notes": "",
    },
    "nativeAmount": "1000",
    "networkFeeOption": "standard",
    "publicAddress": "123123123",
  },
  "pending": false,
  "pin": "",
  "spendInfo": Object {
    "currencyCode": "BTC",
    "metadata": Object {
      "name": "airbitz",
    },
    "nativeAmount": "1000",
    "spendTargets": Array [
      Object {
        "currencyCode": "BTC",
        "nativeAmount": "1000",
        "publicAddress": "123123123",
      },
    ],
  },
  "transaction": null,
  "transactionMetadata": Object {
    "name": "airbitz",
  },
}
`;

exports[`sendConfirmation reducer destination newSpendInfo without name 1`] = `
Object {
  "address": "123123123",
  "authRequired": "none",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "1000",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "1000",
    "networkFeeOption": "standard",
    "publicAddress": "123123123",
  },
  "pending": false,
  "pin": "",
  "spendInfo": Object {
    "currencyCode": "BTC",
    "metadata": Object {},
    "nativeAmount": "1000",
    "spendTargets": Array [
      Object {
        "currencyCode": "BTC",
        "nativeAmount": "1000",
        "publicAddress": "123123123",
      },
    ],
  },
  "transaction": null,
  "transactionMetadata": Object {},
}
`;

exports[`sendConfirmation reducer destination updateTransaction with error 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": [Error],
  "forceUpdateGuiCounter": 1,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "100000",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "100000",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": null,
  "transactionMetadata": null,
}
`;

exports[`sendConfirmation reducer destination updateTransaction with pin error 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": [Error: Incorrect Pin],
  "forceUpdateGuiCounter": 1,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": 0,
    "currencyCode": "BCH",
    "date": 0,
    "nativeAmount": "-681",
    "networkFee": "681",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [
      "123123123",
    ],
    "signedTx": "",
    "txid": "",
  },
  "transactionMetadata": null,
}
`;

exports[`sendConfirmation reducer destination updateTransaction with transaction and legacyAddress 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": null,
  "forceUpdateGuiCounter": 1,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "100000",
  "parsedUri": Object {
    "currencyCode": "BCH",
    "customNetworkFee": Object {},
    "legacyAddress": "123412341234",
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "100000",
    "networkFeeOption": "standard",
    "publicAddress": "bitcoincash:qpltjkre069mp80ylcj87832ju3zt2gr6gercn9j2z",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": 0,
    "currencyCode": "BCH",
    "date": 0,
    "nativeAmount": "-681",
    "networkFee": "681",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [
      "123123123",
    ],
    "signedTx": "",
    "txid": "",
  },
  "transactionMetadata": null,
}
`;

exports[`sendConfirmation reducer destination updateTransaction with transaction and name 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": null,
  "forceUpdateGuiCounter": 1,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "100000",
  "parsedUri": Object {
    "currencyCode": "BCH",
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "airbitz",
      "notes": "",
    },
    "nativeAmount": "100000",
    "networkFeeOption": "standard",
    "publicAddress": "bitcoincash:qpltjkre069mp80ylcj87832ju3zt2gr6gercn9j2z",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": 0,
    "currencyCode": "BCH",
    "date": 0,
    "nativeAmount": "-681",
    "networkFee": "681",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [
      "123123123",
    ],
    "signedTx": "",
    "txid": "",
  },
  "transactionMetadata": Object {
    "name": "airbitz",
  },
}
`;

exports[`sendConfirmation reducer error MAKE_PAYMENT_PROTOCOL_TRANSACTION_FAILED 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": [Error],
  "forceUpdateGuiCounter": 0,
  "isEditable": false,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
  "transactionMetadata": null,
}
`;

exports[`sendConfirmation reducer error UPDATE_TRANSACTION 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": [Error],
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": null,
  "transactionMetadata": null,
}
`;

exports[`sendConfirmation reducer initialState 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
  "transactionMetadata": null,
}
`;

exports[`sendConfirmation reducer isEditable MAKE_PAYMENT_PROTOCOL_TRANSACTION_FAILED 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": [Error],
  "forceUpdateGuiCounter": 0,
  "isEditable": false,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
  "transactionMetadata": null,
}
`;

exports[`sendConfirmation reducer isEditable UPDATE_PAYMENT_PROTOCOL_TRANSACTION 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": false,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "id": "123",
    "nativeAmount": "123",
  },
  "transactionMetadata": null,
}
`;

exports[`sendConfirmation reducer pending 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": true,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
  "transactionMetadata": null,
}
`;

exports[`sendConfirmation reducer pin 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "1234",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
  "transactionMetadata": null,
}
`;

exports[`sendConfirmation reducer reset 1`] = `
Object {
  "address": "",
  "authRequired": "none",
  "error": null,
  "forceUpdateGuiCounter": 0,
  "isEditable": true,
  "isKeyboardVisible": false,
  "nativeAmount": "0",
  "parsedUri": Object {
    "customNetworkFee": Object {},
    "metadata": Object {
      "amountFiat": 0,
      "bizId": 0,
      "category": "",
      "miscJson": "",
      "name": "",
      "notes": "",
    },
    "nativeAmount": "0",
    "networkFeeOption": "standard",
    "publicAddress": "",
  },
  "pending": false,
  "pin": "",
  "spendInfo": null,
  "transaction": Object {
    "blockHeight": -1,
    "currencyCode": "",
    "date": 0,
    "metadata": Object {},
    "nativeAmount": "0",
    "networkFee": "",
    "otherParams": Object {},
    "ourReceiveAddresses": Array [],
    "parentNetworkFee": "",
    "signedTx": "",
    "txid": "",
  },
  "transactionMetadata": null,
}
`;
